File: Practice/22/C++/22/22.cpp
#include "pch.h"
#include <iostream>

/*
Суть алгоритма:
 Лабиринт можно представить в виде дерева, в котором конечными элементами являются тупик или финиш.
 Между ними и началом пути находятся развилки.Таким образом, выход можно найти пробежавшись по всем веткам этого дерева.
*/

std::string maze[25] = {
	"####B######################",
	"# #       #   #      #    #",
	"# # # ###### #### ####### #",
	"# # # #       #   #       #",
	"#   # # ######### # ##### #",
	"# # # #   #       #     # #",
	"### ### ### ############# #",
	"# #   #     # #           #",
	"# # #   ####### ###########",
	"# # # #       # #         C",
	"# # ##### ### # # ####### #",
	"# # #     ### # #       # #",
	"#   ##### ### # ######### #",
	"######### ### #           #",
	"# ####### ### #############",
	"A           #   ###   #   #",
	"# ############# ### # # # #",
	"# ###       # # ### # # # #",
	"# ######### # # ### # # # F",
	"#       ### # #     # # # #",
	"# ######### # ##### # # # #",
	"# #######   #       #   # #",
	"# ####### ######### #######",
	"#         #########       #",
	"#######E############D######"
};

bool checkNear(short direction,short x, short y);

void checkFinish(short x, short y);

void findPath(short direction, short x, short y);

void checkPath(short direction, short x, short y);

void findPath(short direction, short x, short y);

bool finish = false;
int main()
{
	setlocale(LC_ALL,"russian");
	short x, y = 0;
	std::cin >> x >> y;
	if ((x < 0 || x >= 27 || y < 0 || y >= 25) || maze[y].at(x) == '#')
		std::cout << "Не верные координаты";
	else
	{
		checkPath(0,x,y);
		if (!finish)
			std::cout << "Выхода нет";
	}
}
/*Справка по направлениям:

	1
	|
4<-- -->2
	|
	3
*/

//Проверка на тунель(есть только одна свободная точка по направлению)
bool checkNear(short direction,short x, short y)
{
	if (direction == 1 && maze[y - 1].at(x) == ' ' && maze[y].at(x + 1) == '#' && maze[y].at(x - 1) == '#')
		return true;
	else if (direction == 2 && maze[y].at(x + 1) == ' ' && maze[y - 1].at(x) == '#' && maze[y + 1].at(x) == '#')
		return true;
	else if (direction == 3 && maze[y + 1].at(x) == ' ' && maze[y].at(x + 1) == '#' && maze[y].at(x - 1) == '#')
		return true;
	else if (direction == 4 && maze[y].at(x - 1) == ' ' && maze[y - 1].at(x) == '#' && maze[y + 1].at(x) == '#')
		return true;
	else return false;
}

void checkFinish(short x, short y)
{
	for (char i = 'A';i <= 'F';i++)
	{
		if (maze[y - 1].at(x) == i)
		{
			std::cout << i << " ";
			finish = true;
		}
		if (maze[y].at(x + 1) == i)
		{
			std::cout << i << " ";
			finish = true;
		}
		if (maze[y + 1].at(x) == i)
		{
			std::cout << i << " ";
			finish = true;
		}
		if (maze[y].at(x - 1) == i)
		{
			std::cout << i << " ";
			finish = true;
		}
	}
}
//Делаем как Ник
void checkPath(short direction, short x, short y)
{
	if (direction != 3 && maze[y - 1].at(x) == ' ')
		findPath(1, x, y - 1);
	if (direction != 4 && maze[y].at(x + 1) == ' ')
		findPath(2, x + 1, y);
	if (direction != 1 && maze[y + 1].at(x) == ' ')
		findPath(3, x, y + 1);
	if (direction != 2 && maze[y].at(x - 1) == ' ')
		findPath(4, x - 1, y);
}

void findPath(short direction, short x, short y)
{
	while (checkNear(direction,x, y))//пропускаем все точки пока не дойдем до развилки, финиша или тупика
	{
		switch (direction)
		{
		case 1:
			y -= 1;
			break;
		case 2:
			x += 1;
			break;
		case 3:
			y += 1;
			break;
		case 4:
			x -= 1;
			break;
		}
	}
	checkFinish(x, y);
	checkPath(direction, x, y);
}

File: Practice/22/Python/22.py
maze = [
    "####B######################",
    "# #       #   #      #    #",
    "# # # ###### #### ####### #",
    "# # # #       #   #       #",
    "#   # # ######### # ##### #",
    "# # # #   #       #     # #",
    "### ### ### ############# #",
    "# #   #     # #           #",
    "# # #   ####### ###########",
    "# # # #       # #         C",
    "# # ##### ### # # ####### #",
    "# # #     ### # #       # #",
    "#   ##### ### # ######### #",
    "######### ### #           #",
    "# ####### ### #############",
    "A           #   ###   #   #",
    "# ############# ### # # # #",
    "# ###       # # ### # # # #",
    "# ######### # # ### # # # F",
    "#       ### # #     # # # #",
    "# ######### # ##### # # # #",
    "# #######   #       #   # #",
    "# ####### ######### #######",
    "#         #########       #",
    "#######E############D######"
]

finish = False

def checkPath(direction,x,y):
	if (direction != 3 and maze[y - 1][x] == ' '):
		findPath(1, x, y - 1)
	if (direction != 4 and maze[y][x + 1] == ' '):
		findPath(2, x + 1, y)
	if (direction != 1 and maze[y + 1][x] == ' '):
		findPath(3, x, y + 1)
	if (direction != 2 and maze[y][x - 1] == ' '):
		findPath(4, x - 1, y)

def findPath(direction,x,y):
	while checkNear(direction,x, y):
		if direction == 1:
		    y -= 1
		elif direction == 2:
		    x += 1
		elif direction == 3:
		    y += 1
		elif direction == 4:
		    x -= 1
	checkFinish(x, y);
	checkPath(direction, x, y);

def checkNear(direction,x,y):
	if direction == 1 and maze[y - 1][x] == ' ' and maze[y][x + 1] == '#' and maze[y][x - 1] == '#':
		return True
	elif direction == 2 and maze[y][x + 1] == ' ' and maze[y - 1][x] == '#' and maze[y + 1][x] == '#':
		return True
	elif direction == 3 and maze[y + 1][x] == ' ' and maze[y][x + 1] == '#' and maze[y][x - 1] == '#':
		return True
	elif direction == 4 and maze[y][x - 1] == ' ' and maze[y - 1][x] == '#' and maze[y + 1][x] == '#':
		return True
	else:
	    return False

def checkFinish(x,y):
    letters = ('A','B','C','D','E','F')
    for i in letters:
        if maze[y - 1][x] == i:
            print(i,end = " ")
            global finish
            finish = True
        if maze[y][x + 1] == i:
            print(i,end = " ")
            global finish
            finish = True
        if (maze[y + 1][x] == i):
            print(i,end = " ")
            global finish
            finish = True
        if (maze[y][x - 1] == i):
            print(i,end = " ")
            global finish
            finish = True

x, y = map(int,input().split())
if (x < 0 or x >= 27 or y < 0 or y >= 25) or (maze[y][x] == '#'):
	print("Не верные координаты")
else:
	checkPath(0,x,y)
	if finish == False:
		print("Выхода нет")


